File: algorithm/super_ugly_number/go/super_ugly_number.go
package main

import (
	"fmt"
)

func nthUglyNumber(n int, primes []int) int {
	m := make([]int, len(primes))
	copy(m, primes)
	index := make([]int, len(primes))
	seq := make([]int, n)
	seq[0] = 1
	for i := 1; i < n; i++ {
		minV := min(m)
		seq[i] = minV
		for j := 0; j < len(index); j++ {
			if minV == m[j] {
				index[j]++
				m[j] = primes[j] * seq[index[j]]
			}
		}
	}
	return seq[n-1]
}

func min(nums []int) int {
	m := nums[0]
	for i := 1; i < len(nums); i++ {
		if m > nums[i] {
			m = nums[i]
		}
	}
	return m
}

func main() {
	n := 7
	fmt.Printf("%d-th Ugly Number is %d\n", n, nthUglyNumber(n, []int{2, 3, 5}))
}

File: algorithm/super_ugly_number/super_ugly_number.cpp
#include <iostream>
#include <vector>

using std::vector;


int min(vector<int>& nums) {
	int m = nums[0];
	for (unsigned int i = 1; i < nums.size(); i++) {
		if (m > nums[i]) {
			m = nums[i];
		}
	}
	return m;
}

int nthSuperUglyNumber(int n, vector<int>& primes) {
	vector<int> ugly(n,n);
	vector<int> m = primes;
	vector<int> index(primes.size(), 0);
	ugly[0] = 1;
	for (int i = 1; i < n; i++) {
		int minV = min(m);
		ugly[i] = minV;
		for (unsigned int j = 0; j < index.size(); j++) {
			if (minV == m[j]) {
				index[j]++;
				m[j] = primes[j] * ugly[index[j]];
			}
		}
	}
	return ugly[n-1];
}

int main(void) {
	vector<int> primes;
	primes.push_back(2);
	primes.push_back(7);
	primes.push_back(13);
	primes.push_back(19);
	std::cout << nthSuperUglyNumber(12, primes) << std::endl;
	return 0;
}
